### ISSUE TITLE
gtk3: Opening more gvim tabs than the window width accommodates results in gtk warnings

### ISSUE THREAD

**Issue by stellarhopper** [vim/vim#1328]
Open tens of tabs using `gvim -p file1 file2 ... fileN`
If the number of tabs is large enough that the tab bar will overflow, the following warnings appear in the terminal:

```
(gvim:19474): Gtk-WARNING **: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)

(gvim:19474): Gtk-WARNING **: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)

(gvim:19474): Gtk-WARNING **: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)

(gvim:19474): Gtk-WARNING **: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)

(gvim:19474): Gtk-WARNING **: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)

(gvim:19474): Gtk-WARNING **: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)
```

gvim on Fedora 24:

```
$ gvim --version
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Dec  5 2016 19:40:36)
Included patches: 1-124
Modified by <bugzilla@redhat.com>
Compiled by <bugzilla@redhat.com>
Huge version with GTK3 GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_old_static
+arabic          +find_in_path    -mouse_sysmouse  -tag_any_white
+autocmd         +float           +mouse_urxvt     -tcl
+balloon_eval    +folding         +mouse_xterm     +termguicolors
+browse          -footer          +multi_byte      +terminfo
++builtin_terms  +fork()          +multi_lang      +termresponse
+byte_offset     +gettext         -mzscheme        +textobjects
+channel         -hangul_input    +netbeans_intg   +timers
+cindent         +iconv           +num64           +title
+clientserver    +insert_expand   +packages        +toolbar
+clipboard       +job             +path_extra      +user_commands
+cmdline_compl   +jumplist        +perl/dyn        +vertsplit
+cmdline_hist    +keymap          +persistent_undo +virtualedit
+cmdline_info    +lambda          +postscript      +visual
+comments        +langmap         +printer         +visualextra
+conceal         +libcall         +profile         +viminfo
+cryptv          +linebreak       +python/dyn      +vreplace
+cscope          +lispindent      +python3/dyn     +wildignore
+cursorbind      +listcmds        +quickfix        +wildmenu
+cursorshape     +localmap        +reltime         +windows
+dialog_con_gui  +lua/dyn         +rightleft       +writebackup
+diff            +menu            +ruby/dyn        +X11
+digraphs        +mksession       +scrollbind      -xfontset
+dnd             +modify_fname    +signs           +xim
-ebcdic          +mouse           +smartindent     +xpm
+emacs_tags      +mouseshape      +startuptime     +xsmp_interact
+eval            +mouse_dec       +statusline      +xterm_clipboard
+ex_extra        +mouse_gpm       -sun_workshop    -xterm_save
+extra_search    -mouse_jsbterm   +syntax          
+farsi           +mouse_netterm   +tag_binary      
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "/etc/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/etc"
 f-b for $VIMRUNTIME: "/usr/share/vim/vim80"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib64/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/freetype2 -I/usr/include/libdrm -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include   -O2 -g -pipe -Wall -Werror=format-security -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L. -Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,--enable-new-dtags -Wl,-z,relro   -Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -L/usr/local/lib -Wl,--as-needed -o vim   -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lSM -lICE -lXpm -lXt -lX11 -lSM -lICE  -lm -lnsl  -lselinux   -lncurses -lacl -lattr -lgpm -ldl   -Wl,--enable-new-dtags -Wl,-z,relro  -Wl,-z,relro  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib64/perl5/CORE -lperl -lpthread -lresolv -lnsl -ldl -lm -lcrypt -lutil -lc        
```

**Comment by nuko8** [comment]
When gtk applications compute the layout, they get parent widgets to ask their children for their **minimum** and **natural** sizes.

Though I don't see any of the reference manuals giving definition or standards  of them, *minimum sizes* are usually considered to be the ones that are required to keep widgets *usable for normal operations*, and normal sizes are the ones with which (widget designers think) users can work comfortably.

Hence, what you get gvim to do deserves warning, and gtk responds to that appropriately.

**Comment by chrisbra** [comment]
closing then.

**Comment by tonymec** [comment]
@stellarhopper : you might want to open more windows per tab page in such a case, so that you need fewer tab pages to accomodate them all.

Best regards,
Tony.

**Comment by ershov** [comment]
Can you guys reopen this bug?

If this is an intended behavior, there should not be any warnings in console.

**Comment by nuko8** [comment]
> Can you guys reopen this bug?
> 
> If this is an intended behavior, there should not be any warnings in console.

This is not the right place to ask that, because it's the GUI library that sends warnings to the console; there's no code in Vim having gvim do that.

Also, I personally don't think that behavior deserves to be named a bug.  I rather appreciate and prefer that kind of warnings to sudden crash without any warning against insane usage.

**Comment by k-takata** [comment]
I don't think this kind of GTK warnings are useful for end users at all. They are useful only for developers.
Aren't there any ways to disable these warnings?

**Comment by nuko8** [comment]
> I don't think this kind of GTK warnings are useful for end users at all. They are useful only for developers.

I disagree.  Those warnings would eventually help users since they could more or less give us a very first clue to looking into a given issue further, especially when that's a true Vim issue about which we have to do something as soon as possible.  Even if we suppress the warnings, would it be practical or beneficial for us to tell end users who have reported bugs to us, "OK, we got it.  Then could you enable warning output by following some technical instructions we're going to explain to you, then report back the result, please?"   That chore wouldn't bring any benefit to either users or us, as it would take at least hours or days for both to finish.

If they don't want to have any warning from the library, they should launch the GUI from the launch pad, the dock, or whatever the desktop environment offers to them for that purpose.  Or, if they want to launch gvim from the console all the way by typing in the command, they can easily append output redirection to it, say `2>/dev/null`, with only a few extra key strokes.

**Comment by ershov** [comment]
For the purpose of debugging vim has its own logs which can be switched on on demand.
I am a vim user, I do not need to see any warnings to vim developers.

**Comment by nuko8** [comment]
> For the purpose of debugging vim has its own logs which can be switched on on demand.

That won't work, if any, because, I already told you, those warnings come from the library, not from Vim; if they did, they would be marked with `E<number>` in the warning messages; actually, they don't.

> I am a vim user, I do not need to see any warnings to vim developers.

So the warnings are to be delivered equally to every user of the library.  You'll have the warnings if you do anything deserving them.  That has nothing to do with who you are.

**Comment by tonymec** [comment]
@ershov  : IIUC, GTK warnings should appear only in the GUI, or when the GUI is about to be started. If you don't want to see console messages from the GUI (which is understandable), you could start gvim with redirection "> /dev/null 2>&1", or from a desktop icon, couldn't you? Or are there other console messages which you do want to see?

OTOH, you could also log them to some disk file (e.g. ">> ~/vim.log 2>&1"): that way there would be no obnoxious messages to the shell from which you started gvim, but you would still have something to refer to in the case of a crash.

gvim's own error messages (Exxx as mentioned in a previous post) appear at the bottom of the GUI itself after all, not on stdout/stderr.

Best regards,
Tony.

**Comment by ershov** [comment]
@nuko8 the mistake here is to consider the users of **vim** as users of **library**. This is just wrong.
The users of **text editor** are not doing anything deserving a warning.
Vim devs do.
The devs are doing something to the library which is not quite right but eventually is ok. I do not really understand why a library would complain about that on the user's terminal. And definitely these messages are absolutely useless to the end user.

@tonymec yes I do want to see the error messages that relate to my actions or affect my user experience.
Good examples are: bad command line flags, vim unable to connect to X server, etc.
If there were no useful console messages at all, it would make sense to silence stdout and stderr right in gvim and not make users do an additional manual work.

Anyway
As I can see you guys have a strong opinion which you are not going to change.
So I'll just switch to the GTK2 version which does not have that problem.
This is what the user can do about GTK3 warnings.

Cheers,
Yury

**Comment by nuko8** [comment]
> the mistake here is to consider the users of vim as users of library. 

Vim with GTK3 GUI is a client program of GTK+ 3 library; therefore, being a user of that GUI version of the editor is sufficient for being a user of that GUI library.  (I thought that was too obvious to worth giving a proof to that.)

> The devs are doing something to the library which is not quite right but eventually is ok.

That's the user who does something insane, e.g., "Opening more gvim tabs than the window width accommodates."  How on earth could anyone expect the editor to still keep working well under such sadistic madness?  Rather, one should appreciate the robustness of the GUI; thankfully, it won't bring you anything harmful more than the warnings even when it is mercilessly forced to put into a state where letting it abort without warning in order to avoid more disastrous consequences is no surprise at all.  Owing to that, one won't lose any bit of the changes he has done in a session with the GUI. instead of regretting pushing it to the limit for no good reason afterwards.

> I do not really understand why a library would complain about that on the user's terminal. 

So, at the beginning, I said, "This is not the right place to ask that."  Ask the library devs about that, if you will.

> And definitely these messages are absolutely useless to the end user.

Maybe, to some of the end users.

> As I can see you guys have a strong opinion which you are not going to change.

That's rather the way things have been working so far than our arbitrary, strong opinions here.  Sorry, but I need tell you the same thing again, "This is not the right place to ask that."  If you really abhor the warnings, why don't you go to GTK+ project and ask them to change their way?  Needless to say, that implies asking them for change in a huge software suite consisting of glib, cairo, pango, gdk and gtk.

> So I'll just switch to the GTK2 version which does not have that problem.

Obviously, being unaware of any problem there is different from having no problem there.

**Comment by juanlanus** [comment]
This is a usability issue.  
Seeing a raft of warnings after a licit action breaks flow state.  
Seeing the same warnings over and over is annoying and brings up thoughts about maintenance.  
  
My scenario: I open many small (JS) files at once, and store the session with :mks. 
When I restart with -S flag I see all those "gvim 7381" messages: 
(gvim:7381)  Drawing a gadget with negative dimensions 
I don't look at them, after having wasted some time to find out what they mean. 
So now I ignore them in block, possibly overlooking a really meaningful message. 

These are the two issues: having to look at pointless messages, or possibly missing a useful one.  
  
vim is known by its huge usability, no other editor on earth allows us developers to do so much work with such a minimal effort, and it's a shame to see it start as if it was a version 0.1 application.  
  
The high usability thing is because vim has the best ratio of work done vs. keystrokes and mishaps. 
The initial investment to grok how vi works is greatly compensated with how it is a fast and safe tool to edit text. 
Usability is not defined in terms of ease to use, it's about efficiency, efficacy and satisfaction (the ISO definition).

**Comment by dpelle** [comment]
@nuko8  wrote:

> That's the user who does something insane, e.g., "Opening more gvim tabs than the window width accommodates."

There is nothing insane here.  In fact I can reproduce the gtk warning with only 2 tabs:
```
$ vim -f -g -p 1 2
(gvim:19117): Gtk-WARNING **: 23:59:28.337: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)

(gvim:19117): Gtk-WARNING **: 23:59:28.339: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)
```

I'm using the latest vim-8.2.2840 with gtk3 gui on Linux xubuntu-18.04.5.
With other other GUIs (gtk2, Motif, Athena) I do not see any warning when running `vim -f -g p 1 2`.

I do not know gtk3 enough to tell for sure, but it looks like a mis-use of the gtk3 API somewhere in Vim.

**Comment by brammool** [comment]
This requires someone with GTK3 knowledge to figure out if there is a workaround.

**Comment by JakeMercer** [comment]
I got fed up with seeing these warnings (I make heavy use of `-p`) so I hid them with the following in my .bashrc:

```
# Hide all warnings from gvim so we don't get GTK spewing all over the screen.
gvim() {
    command gvim "$@" > /dev/null 2>&1
}
```

**Comment by juanlanus** [comment]
@JakeMercer : Instead of muting the messages, is it possible to pipe them into a filter that would only swallow lines containing "(gvim:7381)"?
